transfer/basic_download.go

```diff
@@ -0,0 +1,242 @@
+	return os.RemoveAll(a.tempDir())
+}
+
+func (a *basicDownloadAdapter) tempDir() string {
+	// Must be dedicated to this adapter as deleted by ClearTempStorage
+	// Also make local to this repo not global, and separate to localstorage temp,
+	// which gets cleared at the end of every invocation
+	d := filepath.Join(localstorage.Objects().RootDir, "incomplete")
+	if err := os.MkdirAll(d, 0755); err != nil {
+		return os.TempDir()
+	}
+	return d
+}
+
+func (a *basicDownloadAdapter) DoTransfer(t *Transfer, cb TransferProgressCallback, authOkFunc func()) error {
+
+	f, fromByte, hashSoFar, err := a.checkResumeDownload(t)
+	if err != nil {
+		return err
+	}
+	return a.download(t, cb, authOkFunc, f, fromByte, hashSoFar)
+}
+
+// Checks to see if a download can be resumed, and if so returns a non-nil locked file, byte start and hash
+func (a *basicDownloadAdapter) checkResumeDownload(t *Transfer) (outFile *os.File, fromByte int64, hashSoFar hash.Hash, e error) {
+	// lock the file by opening it for read/write, rather than checking Stat() etc
+	// which could be subject to race conditions by other processes
+	f, err := os.OpenFile(a.downloadFilename(t), os.O_RDWR, 0644)
+
+	if err != nil {
+		// Create a new file instead, must not already exist or error (permissions / race condition)
+		newfile, err := os.OpenFile(a.downloadFilename(t), os.O_CREATE|os.O_WRONLY|os.O_EXCL, 0644)
+		return newfile, 0, nil, err
+	}
+
+	// Successfully opened an existing file at this point
+	// Read any existing data into hash then return file handle at end
+	hash := tools.NewLfsContentHash()
+	n, err := io.Copy(hash, f)
+	if err != nil {
+		f.Close()
+		return nil, 0, nil, err
+	}
+	tracerx.Printf("xfer: Attempting to resume download of %q from byte %d", t.Object.Oid, n)
+	return f, n, hash, nil
+
+}
+
+// Create or open a download file for resuming
+func (a *basicDownloadAdapter) downloadFilename(t *Transfer) string {
+	// Not a temp file since we will be resuming it
+	return filepath.Join(a.tempDir(), t.Object.Oid+".tmp")
+}
+
+// download starts or resumes and download. Always closes dlFile if non-nil
+func (a *basicDownloadAdapter) download(t *Transfer, cb TransferProgressCallback, authOkFunc func(), dlFile *os.File, fromByte int64, hash hash.Hash) error {
+
+	if dlFile != nil {
+		// ensure we always close dlFile. Note that this does not conflict with the
+		// early close below, as close is idempotent.
+		defer dlFile.Close()
+	}
+
+	rel, ok := t.Object.Rel("download")
+	if !ok {
+		return errors.New("Object not found on the server.")
+	}
+
+	req, err := httputil.NewHttpRequest("GET", rel.Href, rel.Header)
+	if err != nil {
+		return err
+	}
+
+	if fromByte > 0 {
+		if dlFile == nil || hash == nil {
+			return fmt.Errorf("Cannot restart %v from %d without a file & hash", t.Object.Oid, fromByte)
+		}
+		// We could just use a start byte, but since we know the length be specific
+		req.Header.Set("Range", fmt.Sprintf("bytes=%d-%d", fromByte, t.Object.Size))
```

Comment:
Should this be `t.Object.Size -1` ?
RFC 7233 seems to suggest the range is zero based inclusive. See RFC 7233 section 2.1.